src/distributions/integer.rs

```diff
@@ -109,53 +114,63 @@ impl_nzint!(NonZeroU64, NonZeroU64::new);
 impl_nzint!(NonZeroU128, NonZeroU128::new);
 impl_nzint!(NonZeroUsize, NonZeroUsize::new);
 
-#[cfg(feature = "simd_support")]
-macro_rules! simd_impl {
-    ($(($intrinsic:ident, $vec:ty),)+) => {$(
+macro_rules! intrinsic_impl {
+    ($($intrinsic:ident),+) => {$(
+        /// Available only on x86/64 platforms
         impl Distribution<$intrinsic> for Standard {
             #[inline]
             fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $intrinsic {
-                $intrinsic::from_bits(rng.gen::<$vec>())
+                // On proper hardware, this should compile to SIMD instructions
+                // Verified on x86 Haswell with __m128i, __m256i
+                let mut buf = [0_u8; mem::size_of::<$intrinsic>()];
+                rng.fill_bytes(&mut buf);
+                // x86 is little endian so no need for conversion
+                // SAFETY: we know [u8; N] and $intrinsic have the same size
+                unsafe { mem::transmute_copy(&buf) }
             }
         }
     )+};
+}
 
-    ($bits:expr,) => {};
-    ($bits:expr, $ty:ty, $($ty_more:ty,)*) => {
-        simd_impl!($bits, $($ty_more,)*);
-
+#[cfg(feature = "simd_support")]
+macro_rules! simd_impl {
+    ($($ty:ty),+) => {$(
+        /// Requires nightly Rust and the [`simd_support`] feature
+        ///
+        /// [`simd_support`]: https://github.com/rust-random/rand#crate-features
         impl Distribution<$ty> for Standard {
             #[inline]
             fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
-                let mut vec: $ty = Default::default();
-                unsafe {
-                    let ptr = &mut vec;
-                    let b_ptr = &mut *(ptr as *mut $ty as *mut [u8; $bits/8]);
-                    rng.fill_bytes(b_ptr);
+                // TODO: impl this generically once const generics are robust enough
+                let mut vec: Simd<u8, { mem::size_of::<$ty>() }> = Default::default();
+                rng.fill_bytes(vec.as_mut_array());
+                // NOTE: replace with `to_le` if added to core::simd
+                #[cfg(not(target_endian = "little"))]
+                {
+                    vec = vec.reverse();
                 }
-                vec.to_le()
+                // SAFETY: we know u8xN and $ty have the same size
+                unsafe { mem::transmute_copy(&vec) }
```

Comment:
If you use `Fill` (with the correct element type) it will do the Endianness conversion here. `reverse` is not the correct conversion IIUC.

Reply:
Oh nice that'll also make it more generic